File: Algorithms/Python3/Binary_Search.py
'''
    FUNCTION binary-search(A,low,high,x):
        while low <= high do
            mid = (low + high) /2
            if x = A[mid] then
                return mid
            else if x > A[mid] then
                low = low + 1
            else
                high = high - 1
        return nil
'''
def binary_search(a, low, high, x):
    while low <= high:
        mid = (low + high) // 2
        if x == a[mid]: return mid
        elif x > a[mid]: low += 1
        else: high -= 1
    return -1

a = [1, 2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10]
low = 2
high = 8 

bs = binary_search(a, low, high, 4)

if bs != -1: print(4," is at position ",bs)

File: Algorithms/Cpp/Binary_Search.cpp
#include <vector>
#include <iostream>
using namespace std;

int binary_search(vector<int> a, int low, int high, int x){

    while (low <= high){
        int mid = (low+high)/2;
        if (a.at(mid) == x)
            return mid;
        else if (x > a.at(mid))
            low = mid + 1;
        else
            high = mid - 1;
    }

    return -1;
}


int main() {

    vector<int> a = {1,2,3,4,5,6,7,8,9,10};
    int low = 2;
    int high = 9;

    int bs = binary_search(a,low,high,4);

    cout << "4 is at position: " << bs << endl;
    
    return 0;
}